src/tsolvers/lasolver/Simplex.cpp

```diff
@@ -0,0 +1,490 @@
+            return val >= bound.getValue();
+        }
+    }
+}
+
+bool Simplex::checkSimplex(std::vector<LABoundRef> &explanation, std::vector<opensmt::Real> &explanationCoefficients) {
+
+    explanation.clear( );
+    explanationCoefficients.clear( );
+
+    bool bland_rule = false;
+    unsigned repeats = 0;
+    unsigned pivot_counter = 0;
+    unsigned bland_counter = 0;
+    // These values are from Yices
+    unsigned bthreshold = bland_threshold;
+    if (boundStore.nVars() > 10000)
+        bthreshold *= 1000;
+    else if (boundStore.nVars() > 1000)
+        bthreshold *= 100;
+
+    // keep doing pivotAndUpdate until the SAT/UNSAT status is confirmed
+    while (true) {
+        repeats++;
+        LVRef x = LVRef_Undef;
+
+        if (!bland_rule && (repeats > tableau.getNumOfCols()))
+            bland_rule = true;
+
+        if (bland_rule) {
+            x = getBasicVarToFixByBland();
+            ++bland_counter;
+            ++simplex_stats.num_bland_ops;
+        }
+        else {
+            x = getBasicVarToFixByShortestPoly();
+            ++pivot_counter;
+            ++simplex_stats.num_pivot_ops;
+        }
+
+        if (x == LVRef_Undef) {
+            // SAT
+            refineBounds();
+            return true;
+        }
+
+        LVRef y_found = LVRef_Undef;
+        if (bland_rule){
+            y_found = findNonBasicForPivotByBland(x);
+        }
+        else{
+            y_found = findNonBasicForPivotByHeuristic(x);
+        }
+        // if it was not found - UNSAT
+        if (y_found == LVRef_Undef) {
+            getConflictingBounds(x, explanation, explanationCoefficients);
+            return false;
+        }
+            // if it was found - pivot old Basic x with non-basic y and do the model updates
+        else {
+            pivot(x, y_found);
+        }
+    }
+}
+
+bool Simplex::isModelOutOfBounds(LVRef v) const {
+    return ( (model.read(v) > model.Ub(v)) || (model.read(v) < model.Lb(v)) );
+}
+
+bool Simplex::isModelOutOfUpperBound(LVRef v) const
+{
+    return ( model.read(v)> model.Ub(v) );
+}
+
+bool Simplex::isModelOutOfLowerBound(LVRef v) const
+{
+    return ( model.read(v) < model.Lb(v) );
+}
+
+const Delta Simplex::overBound(LVRef v) const
+{
+    assert( isModelOutOfBounds(v) );
+    if (isModelOutOfUpperBound(v))
+    {
+        return ( Delta(model.read(v) - model.Ub(v)) );
+    }
+    else if ( isModelOutOfLowerBound(v) )
+    {
+        return ( Delta(model.Lb(v) - model.read(v)) );
+    }
+    assert (false);
+    printf("Problem in overBound, LRASolver.C:%d\n", __LINE__);
+    exit(1);
+}
+
+bool Simplex::isEquality(LVRef v) const
+{
+    return model.isEquality(v);
+}
+
+bool Simplex::isUnbounded(LVRef v) const
+{
+    bool rval = model.isUnbounded(v);
+//    if (rval)
+//        printf("Var %s is unbounded\n", lva.printVar(v));
+    return rval;
+}
+
+std::unique_ptr<Polynomial> Simplex::addPoly(std::unique_ptr<std::vector<std::pair<LVRef,opensmt::Real>>> terms) {
```

Comment:
This method seems useless.
The vector of terms passed here should never have been created.
Instead the Polynomial should be created directly.

I believe the only important thing is the call to tableau.nonbasicVar
There should be a way to do that directly.

Reply:
That's probably a good idea!